src/coreclr/jit/importer.cpp

```diff
@@ -4167,15 +4169,87 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
         {
+            JITDUMP("checking if we can do anything for a large struct ...");
+            const int MaxStructSize = 16;
+            if ((totalSize == 0) || (totalSize > MaxStructSize))
+            {
+                // Limit to 16 bytes for better throughput, larger structs didn't increase the diffs
+                JITDUMP("struct is larger than 16 bytes - bail out.");
+                return nullptr;
+            }
+
+            uint8_t buffer[MaxStructSize] = {0};
+            if (info.compCompHnd->getReadonlyStaticFieldValue(field, buffer, totalSize))
+            {
+                for (unsigned i = 0; i < totalSize; i++)
+                {
+                    if (buffer[i] != 0)
+                    {
+                        // Value is not all zeroes - bail out
+                        JITDUMP("value is not all zeros - bail out.");
+                        return nullptr;
+                    }
+                }
+
+                DWORD typeFlags = info.compCompHnd->getClassAttribs(fieldClsHnd);
+                if (StructHasOverlappingFields(typeFlags) || StructHasCustomLayout(typeFlags))
+                {
+                    // Technically, we only care about holes in a struct, but overlapping fields would complicate
+                    // the holes detection logic.
+                    JITDUMP("struct has complex layout - bail out.");
+                    return nullptr;
+                }
+
+                unsigned totalFldSize = 0;
+                for (unsigned fldIndex = 0; fldIndex < fieldsCnt; fldIndex++)
+                {
+                    CORINFO_FIELD_HANDLE innerField = info.compCompHnd->getFieldInClass(fieldClsHnd, fldIndex);
+                    CORINFO_CLASS_HANDLE innerFieldClsHnd;
+                    var_types            fieldVarType =
+                        JITtype2varType(info.compCompHnd->getFieldType(innerField, &innerFieldClsHnd, fieldClsHnd));
+
+                    if (!varTypeIsIntegral(fieldVarType))
+                    {
+                        // A conservative check since a proper one is more complicated and doesn't bring new diffs
+                        JITDUMP("struct has non-primitive fields - bail out.");
+                        return nullptr;
+                    }
+                    totalFldSize += genTypeSize(fieldVarType);
+                }
+
+                if (totalFldSize != totalSize)
+                {
+                    JITDUMP("struct has holes - bail out.");
+                    return nullptr;
+                }
+
+                JITDUMP("success! Optimizing to ASG(struct, 0).");
+
+                unsigned structTempNum = lvaGrabTemp(true DEBUGARG("folding static ro fld empty struct"));
+                lvaSetStruct(structTempNum, fieldClsHnd, false);
+                GenTreeLclVar* tree = impCreateLocalNode(structTempNum DEBUGARG(0));
+                impAppendTree(tree, CHECK_SPILL_NONE, impCurStmtDI);
```

Comment:
> `impAppendTree(tree, CHECK_SPILL_NONE, impCurStmtDI)`;

This appears to append an unused `LCL_VAR` to the statement list, or am I misreading the code?

Reply:
Ah, yes, it's a leftover from my initial impl where I wanted to use GT_CNS_VEC for non-zero guids 🙂